### ISSUE TITLE
More examples for twisted.mail (servers)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4609]
|<img alt="pepijn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| pepijn reported|
|-|-|
|Trac ID|trac#4609|
|Type|enhancement|
|Created|2010-07-21 19:02:51Z|

I am currently trying to write a SMTP server using Twisted. Since not all of twisted.mail is documented, the examples that are there where a huge help.

http://twistedmatrix.com/documents/current/mail/examples/

The problem is that the only example for the server side is a barebones SMTP server. It would be really helpful if there where more examples about other basic ingredients every serious SMTP server needs. These include:

- authentication

- SSL/TSL

- extensions and commands

I have attached an example based of the current SMTP example implementing PLAIN and LOGIN authentication. Other examples appreciated.

Attachments:

* [emailserver.tac](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e97/e970b3cc079e69dffe5d63f2e059e57855e1b976/3e307faafd93d3f5743c57f3109ef37c38ca7ff1.tac) (2689 bytes) - added by pepijn on 2010-07-21 19:03:26Z - ESMTP server with PLAIN and LOGIN authentication

<details><summary>Searchable metadata</summary>

```
trac-id__4609 4609
type__enhancement enhancement
reporter__pepijn pepijn
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1279738971000000 1279738971000000
changetime__1283571087000000 1283571087000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_1"></a>|@khorn commented|
|-|-|

If someone were to add examples of the various things suggested in the ticket description, it seems like it would be fairly straightforward to turn the "series" of examples into a full-blown mail server tutorial (though it might be better to handle that on a separate ticket...like say [#3324](https://github.com/twisted/twisted/issues/3324))

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)|@khorn removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

I guess the attached emailserver.tac duplicates a lot of code from the existing emailserver.tac, not sure if and how we want to do something about that without making things more complicated.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @exarkun|
|-|-|

If we treat the attached as a replacement version of emailserver.tac, I think things look okay.  It's a strict improvement, as far as I can tell.  In the future, please submit patches instead of entire modified files though! :)

I do, however, agree with khorn about having a real series of examples of a mail server.  I think it would be nice to do this in the style of web in 60 seconds.  However, we can improve this email example as an intermediate step towards that goal.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [29965]) Integrate auth changes to emailserver.tac example

Author: pepijn
Reviewer: exarkun
Fixes: [#4609](https://github.com/twisted/twisted/issues/4609)

Add authentication using `twisted.cred` to the Twisted Mail email server
example.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
